code/datums/exonet.dm

```diff
@@ -0,0 +1,226 @@
+	var/login = "[desired_name]@[domain]"
+	// It is VERY unlikely that we'll have two players, in the same round, with the same name and branch, but still, this is here.
+	// If such conflict is encountered, a random number will be appended to the email address. If this fails too, no email account will be created.
+	if(find_email_by_name(login))
+		login = "[desired_name][random_id(/datum/computer_file/data/email_account/, 100, 999)]@[domain]"
+	// If even fallback login generation failed, just don't give them an email. The chance of this happening is astronomically low.
+	if(find_email_by_name(login))
+		to_chat(user, "You were not assigned an email address.")
+		user.StoreMemory("You were not assigned an email address.", /decl/memory_options/system)
+	else
+		var/datum/computer_file/data/email_account/EA = new/datum/computer_file/data/email_account(login, user.real_name, assignment)
+		EA.password = GenerateKey()
+		if(user.mind)
+			user.mind.initial_email_login["login"] = EA.login
+			user.mind.initial_email_login["password"] = EA.password
+			user.StoreMemory("Your email account address is [EA.login] and the password is [EA.password].", /decl/memory_options/system)
+		if(issilicon(user))
+			var/mob/living/silicon/S = user
+			var/datum/nano_module/email_client/my_client = S.get_subsystem_from_path(/datum/nano_module/email_client)
+			if(my_client)
+				my_client.stored_login = EA.login
+				my_client.stored_password = EA.password
+
+/datum/exonet/proc/find_email_by_name(var/login)
+	for(var/datum/computer_file/data/email_account/A in email_accounts)
+		if(A.login == login)
+			return A
+	return 0
+
+/datum/exonet/proc/is_connected_plexus()
+	for(var/obj/machinery/computer/exonet/uplink/modem in modems)
+		if(modem.operable())
+			return TRUE
+	return FALSE
+
+/datum/exonet/proc/add_device(var/device, var/keydata)
+	if(!router)
+		return 0 // Uh?? No router? Guess the network is busted.
+	if(router.keydata != keydata)
+		return 0 // Authentication failed.
+
+	if(istype(device, /obj/machinery/computer/exonet/mainframe))
+		mainframes |= device
+	else if(istype(device, /obj/machinery/computer/exonet/broadcaster))
+		broadcasters |= device
+	else if(istype(device, /obj/machinery/computer/exonet/uplink))
+		modems |= device
+	else if(istype(device, /obj/machinery/computer/exonet/broadcaster/router) && !router)
+		router = device // Special setty-uppy-timy for routers.
+	network_devices |= device
+
+/datum/exonet/proc/remove_device(var/device)
+	if(istype(device, /obj/machinery/computer/exonet/mainframe))
+		mainframes -= device
+	else if(istype(device, /obj/machinery/computer/exonet/broadcaster))
+		broadcasters -= device
+	else if(istype(device, /obj/machinery/computer/exonet/uplink))
+		modems -= device
+	network_devices -= device
+
+/datum/exonet/proc/set_router(var/device)
+	if(router)
+		// THERE CAN BE ONLY ONE!!!
+		LAZYREMOVE(network_devices, router)
+		LAZYREMOVE(broadcasters, router)
+	router = device
+	LAZYADD(network_devices, device)
```

Comment:
maybe LAZYDISTINCTADD here and for broadcasters?
I didn't read to ther spot where it's used, but would make it safe to set router that's already in the network.